File: src/main/java/com/hpc/jobagent/service/AnalysisService.java
package com.hpc.jobagent.service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.concurrent.Executor;

import com.hpc.jobagent.agent.AgentReport;
import com.hpc.jobagent.agent.AgentWorkflowService;
import com.hpc.jobagent.config.AgentProperties;
import com.hpc.jobagent.domain.AnalysisCache;
import com.hpc.jobagent.domain.AnalysisStatus;
import com.hpc.jobagent.domain.AnalysisTask;
import com.hpc.jobagent.domain.JobPosting;
import com.hpc.jobagent.domain.ResumeProfile;
import com.hpc.jobagent.dto.AnalysisResponse;
import com.hpc.jobagent.dto.CreateAnalysisRequest;
import com.hpc.jobagent.mapper.AnalysisTaskMapper;
import com.hpc.jobagent.support.NotFoundException;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

@Service
public class AnalysisService {

    private final AnalysisTaskMapper analysisMapper;
    private final ResumeService resumeService;
    private final JobPostingService jobPostingService;
    private final AgentWorkflowService agentWorkflowService;
    private final AnalysisCacheService cacheService;
    private final HashService hashService;
    private final JsonService jsonService;
    private final AgentProperties properties;
    private final Executor analysisExecutor;
    private final CurrentUserService currentUserService;

    public AnalysisService(AnalysisTaskMapper analysisMapper,
                           ResumeService resumeService,
                           JobPostingService jobPostingService,
                           AgentWorkflowService agentWorkflowService,
                           AnalysisCacheService cacheService,
                           HashService hashService,
                           JsonService jsonService,
                           AgentProperties properties,
                           @Qualifier("analysisExecutor") Executor analysisExecutor,
                           CurrentUserService currentUserService) {
        this.analysisMapper = analysisMapper;
        this.resumeService = resumeService;
        this.jobPostingService = jobPostingService;
        this.agentWorkflowService = agentWorkflowService;
        this.cacheService = cacheService;
        this.hashService = hashService;
        this.jsonService = jsonService;
        this.properties = properties;
        this.analysisExecutor = analysisExecutor;
        this.currentUserService = currentUserService;
    }

    public AnalysisResponse createAndRun(CreateAnalysisRequest request) {
        Long userId = currentUserService.userId();
        ResumeProfile resume = resumeService.getEntity(request.resumeId());
        JobPosting job = jobPostingService.getEntity(request.jobId());
        LocalDateTime now = LocalDateTime.now();
        AnalysisTask task = new AnalysisTask();
        task.setUserId(userId);
        task.setResumeId(resume.getId());
        task.setJobId(job.getId());
        task.setStatus(AnalysisStatus.PENDING);
        task.setCacheKey(cacheKey(resume, job));
        task.setCreatedAt(now);
        task.setUpdatedAt(now);
        analysisMapper.insert(task);
        analysisExecutor.execute(() -> runInternal(task.getId()));
        return AnalysisResponse.from(task);
    }

    public AnalysisResponse get(Long id) {
        return AnalysisResponse.from(getEntity(id));
    }

    public List<AnalysisResponse> search(Long resumeId, Long jobId, AnalysisStatus status, int limit) {
        int safeLimit = Math.max(1, Math.min(limit, 100));
        return analysisMapper.search(resumeId, jobId, status, safeLimit, currentUserService.userId()).stream()
                .map(AnalysisResponse::from)
                .toList();
    }

    public void delete(Long id) {
        getEntity(id);
        analysisMapper.deleteByIdAndUserId(id, currentUserService.userId());
    }

    public AnalysisTask getEntity(Long id) {
        AnalysisTask task = analysisMapper.findByIdAndUserId(id, currentUserService.userId());
        if (task == null) {
            throw new NotFoundException("分析任务不存在：" + id);
        }
        return task;
    }

    public void run(Long id) {
        getEntity(id);
        runInternal(id);
    }

    private void runInternal(Long id) {
        AnalysisTask task = analysisMapper.findById(id);
        if (task == null) {
            return;
        }
        LocalDateTime now = LocalDateTime.now();
        analysisMapper.updateStatus(id, AnalysisStatus.RUNNING, null, now);
        try {
            if (properties.getCache().isEnabled()) {
                AnalysisCache cache = cacheService.find(task.getCacheKey());
                if (cache != null) {
                    completeFromCache(task, cache);
                    return;
                }
            }

            ResumeProfile resume = resumeService.getEntityForUser(task.getResumeId(), task.getUserId());
            JobPosting job = jobPostingService.getEntityForUser(task.getJobId(), task.getUserId());
            AgentReport report = agentWorkflowService.run(resume, job);
            completeFromReport(task, report);
        } catch (Exception ex) {
            analysisMapper.updateStatus(id, AnalysisStatus.FAILED, rootMessage(ex), LocalDateTime.now());
        }
    }

    private void completeFromReport(AnalysisTask task, AgentReport report) {
        LocalDateTime now = LocalDateTime.now();
        task.setStatus(AnalysisStatus.COMPLETED);
        task.setMatchScore(report.matchDetail().score());
        task.setSummary(report.summary());
        task.setResumeInsight(jsonService.toJson(report.resumeInsight()));
        task.setJobInsight(jsonService.toJson(report.jobInsight()));
        task.setMatchDetail(jsonService.toJson(report.matchDetail()));
        task.setSuggestions(jsonService.toJson(report.suggestions()));
        task.setInterviewQuestions(jsonService.toJson(report.interviewQuestions()));
        task.setAgentTrace(jsonService.toJson(report.trace()));
        task.setUpdatedAt(now);
        task.setCompletedAt(now);
        analysisMapper.markCompleted(task);
        saveCache(task);
    }

    private void completeFromCache(AnalysisTask task, AnalysisCache cache) {
        LocalDateTime now = LocalDateTime.now();
        task.setStatus(AnalysisStatus.COMPLETED);
        task.setMatchScore(cache.getMatchScore());
        task.setSummary(cache.getSummary() + "（命中缓存）");
        task.setResumeInsight(cache.getResumeInsight());
        task.setJobInsight(cache.getJobInsight());
        task.setMatchDetail(cache.getMatchDetail());
        task.setSuggestions(cache.getSuggestions());
        task.setInterviewQuestions(cache.getInterviewQuestions());
        task.setAgentTrace(cache.getAgentTrace());
        task.setUpdatedAt(now);
        task.setCompletedAt(now);
        analysisMapper.markCompleted(task);
    }

    private void saveCache(AnalysisTask task) {
        if (!properties.getCache().isEnabled()) {
            return;
        }
        AnalysisCache cache = new AnalysisCache();
        cache.setCacheKey(task.getCacheKey());
        cache.setMatchScore(task.getMatchScore());
        cache.setSummary(task.getSummary());
        cache.setResumeInsight(task.getResumeInsight());
        cache.setJobInsight(task.getJobInsight());
        cache.setMatchDetail(task.getMatchDetail());
        cache.setSuggestions(task.getSuggestions());
        cache.setInterviewQuestions(task.getInterviewQuestions());
        cache.setAgentTrace(task.getAgentTrace());
        cacheService.save(cache);
    }

    private String cacheKey(ResumeProfile resume, JobPosting job) {
        return hashService.sha256(resume.getUserId() + "\n---RESUME---\n"
                + resume.getContent() + "\n---JOB---\n" + job.getDescription());
    }

    private String rootMessage(Exception ex) {
        Throwable cursor = ex;
        while (cursor.getCause() != null) {
            cursor = cursor.getCause();
        }
        return cursor.getClass().getSimpleName() + ": " + cursor.getMessage();
    }
}

File: ai-job-python-agent/app/services/analysis_service.py
import hashlib
import json

from fastapi import HTTPException
from sqlalchemy.orm import Session

from app.agents.workflow import agent_workflow
from app.db.session import SessionLocal
from app.models.analysis import AnalysisTask
from app.models.job import JobPosting
from app.models.resume import ResumeProfile
from app.schemas.analysis import AnalysisResponse, CreateAnalysisRequest
from app.services.cache_service import AnalysisCacheService
from app.services.job_service import JobService
from app.services.resume_service import ResumeService
from app.utils.json import dump_json, load_json_object
from app.utils.time import now


class AnalysisService:
    def __init__(self, db: Session) -> None:
        self.db = db

    def create(self, user_id: int, request: CreateAnalysisRequest) -> AnalysisResponse:
        resume = ResumeService(self.db).get_entity(user_id, request.resume_id)
        job = JobService(self.db).get_entity(user_id, request.job_id)
        timestamp = now()
        task = AnalysisTask(
            user_id=user_id,
            resume_id=resume.id,
            job_id=job.id,
            status="PENDING",
            cache_key=self._cache_key(user_id, resume, job),
            created_at=timestamp,
            updated_at=timestamp,
        )
        self.db.add(task)
        self.db.commit()
        self.db.refresh(task)
        return self._response(task)

    def list_by_user(self, user_id: int, status_filter: str | None, limit: int) -> list[AnalysisResponse]:
        query = self.db.query(AnalysisTask).filter(AnalysisTask.user_id == user_id)
        if status_filter:
            query = query.filter(AnalysisTask.status == status_filter.upper())
        tasks = query.order_by(AnalysisTask.id.desc()).limit(max(1, min(limit, 100))).all()
        return [self._response(task) for task in tasks]

    def get_entity(self, user_id: int, analysis_id: int) -> AnalysisTask:
        task = self.db.query(AnalysisTask).filter(AnalysisTask.id == analysis_id, AnalysisTask.user_id == user_id).first()
        if task is None:
            raise HTTPException(status_code=404, detail="analysis task not found")
        return task

    def get_response(self, user_id: int, analysis_id: int) -> AnalysisResponse:
        return self._response(self.get_entity(user_id, analysis_id))

    def mark_pending(self, user_id: int, analysis_id: int) -> AnalysisResponse:
        task = self.get_entity(user_id, analysis_id)
        task.status = "PENDING"
        task.error_message = None
        task.updated_at = now()
        self.db.commit()
        self.db.refresh(task)
        return self._response(task)

    def _cache_key(self, user_id: int, resume: ResumeProfile, job: JobPosting) -> str:
        raw = f"{user_id}\n---RESUME---\n{resume.content}\n---JOB---\n{job.description}"
        return hashlib.sha256(raw.encode()).hexdigest()

    def _response(self, task: AnalysisTask) -> AnalysisResponse:
        return AnalysisResponse(
            id=task.id,
            resume_id=task.resume_id,
            job_id=task.job_id,
            status=task.status,
            match_score=task.match_score,
            summary=task.summary,
            resume_insight=load_json_object(task.resume_insight),
            job_insight=load_json_object(task.job_insight),
            match_detail=load_json_object(task.match_detail),
            suggestions=load_json_object(task.suggestions),
            interview_questions=load_json_object(task.interview_questions),
            agent_trace=[] if task.agent_trace is None else json.loads(task.agent_trace),
            error_message=task.error_message,
            created_at=task.created_at,
            updated_at=task.updated_at,
            completed_at=task.completed_at,
        )


def run_analysis_task(task_id: int) -> None:
    db = SessionLocal()
    try:
        task = db.get(AnalysisTask, task_id)
        if task is None:
            return
        task.status = "RUNNING"
        task.updated_at = now()
        db.commit()

        cache_service = AnalysisCacheService(db)
        cached = cache_service.find(task.cache_key)
        if cached:
            _complete_from_cache(db, task, cached)
            return

        resume = db.query(ResumeProfile).filter(ResumeProfile.id == task.resume_id, ResumeProfile.user_id == task.user_id).first()
        job = db.query(JobPosting).filter(JobPosting.id == task.job_id, JobPosting.user_id == task.user_id).first()
        if resume is None or job is None:
            raise RuntimeError("resume or job not found")

        report = agent_workflow.run(resume, job)
        _complete_from_report(db, task, report)
        cache_service.save(task.cache_key, report)
    except Exception as exc:
        task = db.get(AnalysisTask, task_id)
        if task is not None:
            task.status = "FAILED"
            task.error_message = f"{exc.__class__.__name__}: {exc}"
            task.updated_at = now()
            db.commit()
    finally:
        db.close()


def _complete_from_cache(db: Session, task: AnalysisTask, cached: dict) -> None:
    timestamp = now()
    task.status = "COMPLETED"
    task.match_score = int(cached["match_score"])
    task.summary = str(cached["summary"]) + "（命中缓存）"
    task.resume_insight = cached["resume_insight"]
    task.job_insight = cached["job_insight"]
    task.match_detail = cached["match_detail"]
    task.suggestions = cached["suggestions"]
    task.interview_questions = cached["interview_questions"]
    task.agent_trace = cached["agent_trace"]
    task.error_message = None
    task.updated_at = timestamp
    task.completed_at = timestamp
    db.commit()


def _complete_from_report(db: Session, task: AnalysisTask, report: dict) -> None:
    timestamp = now()
    task.status = "COMPLETED"
    task.match_score = int(report["match_detail"]["score"])
    task.summary = report["summary"]
    task.resume_insight = dump_json(report["resume_insight"])
    task.job_insight = dump_json(report["job_insight"])
    task.match_detail = dump_json(report["match_detail"])
    task.suggestions = dump_json(report["suggestions"])
    task.interview_questions = dump_json(report["interview_questions"])
    task.agent_trace = dump_json(report["agent_trace"])
    task.error_message = None
    task.updated_at = timestamp
    task.completed_at = timestamp
    db.commit()
